content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,433 @@
+
+It’s not a problem in most cases and the possible solutions have big trade offs
+so we are not rushing into fixing it. If it does affect you, please post your
+feedback on the [tracking issue](https://github.com/llvm/llvm-project/issues/72856).
+
+# TableGen Language Server
+
+The MLIR project has implemented a server for the
+[Language Server Protocol (LSP)](https://microsoft.github.io/language-server-protocol/).
+Which supports TableGen and
+[2 other languages](https://mlir.llvm.org/docs/Tools/MLIRLSP/) used within MLIR.
+
+The language server protocol is an editor agnostic means of providing
+information to an editor about the structure of a language and project. For
+instance, where are the include files, where is the definition of this type? If
+you have used an editor like Visual Studio Code, you’ve probably used this
+without knowing it. “Go To Definition” is the classic use case.
+
+It means you can open a project, go to the code you want to change and jump
+directly to the relevant parts of the repository. With 500,000+ lines of
+TableGen in the LLVM Project, that’s a lot of code you get to ignore!
+
+# Setup
+
+You will need a copy of the server binary `tblgen-lsp-server`. Which you can get
+from the
+[release package](https://github.com/llvm/llvm-project/releases) for your
+platform, or you can build it yourself.
+
+This is the minimal set of commands to build it yourself:
+```
+$ cmake -G Ninja ../llvm-project/llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_PROJECTS="mlir"
+$ ninja tblgen-lsp-server
+```
+
+You will find `tblgen-lsp-server` in `<build-dir>/bin/`.
+
+The server reads a compilation database file, which is made for you when you
+configure LLVM using CMake.
+
+As long as your checkout of llvm-project includes
+[this commit](https://github.com/llvm/llvm-project/commit/c4afeccdd235a282d200c450e06a730504a66a08)
+the compilation database will include all TableGen files from all enabled
+projects (previously it was MLIR only).
+
+For example this configure will include information about TableGen files from
+llvm, clang, mlir and lldb:
+```
+$ cmake -G Ninja ../llvm-project/llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_PROJECTS="clang;llvm;lldb;mlir"
+```
+
+This also applies to `-DLLVM_TARGETS_TO_BUILD=`. Enabling only one target will
+mean the compilation database only has files relevant to that target.
+
+**Note:** You do not need to build a project to include its TableGen files in
+the compilation database. Configuring is all that’s needed.
+
+Next, install the MLIR
+[extension](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-mlir)
+into Visual Studio Code (the language server can work with other but setup will
+differ).
+
+Then follow the setup instructions
+[here](https://mlir.llvm.org/docs/Tools/MLIRLSP/#td---tablegen-files) to tell
+the extension where the server and compilation database are.
+
+## Example
+
+Assuming you configured LLVM with the AArch64 target enabled (it is on by
```

Comment:
This example assumes you have configured LLVM with the AArch64 target enabled. (It is enabled by default)